prbench/src/prbench/envs/geom3d/packing3d.py

```diff
@@ -0,0 +1,765 @@
+            )
+        )
+
+    def get_object_triangle_features(
+        self, name: str
+    ) -> tuple[float, float, float, float]:
+        """Get the triangle features (side_a, side_b, depth, triangle_type) of a
+        triangle object."""
+        obj = self.get_object_from_name(name)
+        return (
+            self.get(obj, "side_a"),
+            self.get(obj, "side_b"),
+            self.get(obj, "depth"),
+            self.get(obj, "triangle_type"),
+        )
+
+    @property
+    def objects(self) -> list[Object]:
+        """Get all objects in the state."""
+        objects = list(self.data.keys())
+        return objects
+
+    @property
+    def rack_half_extents(self) -> tuple[float, float, float]:
+        """Get the half extents of the rack."""
+        return self.get_object_half_extents("rack")
+
+    @property
+    def rack_pose(self) -> Pose:
+        """Get the pose of the rack."""
+        return self.get_object_pose("rack")
+
+    @property
+    def part_poses(self) -> dict[str, Pose]:
+        """Get the poses of all parts."""
+        poses = {}
+        for obj in self.objects:
+            if obj.name.startswith("part"):
+                poses[obj.name] = self.get_object_pose(obj.name)
+        return poses
+
+    @property
+    def part_types(self) -> dict[str, Type]:
+        """Get the types of all parts."""
+        types = {}
+        for obj in self.objects:
+            if obj.name.startswith("part"):
+                types[obj.name] = obj.type
+        return types
+
+    @property
+    def part_features(
+        self,
+    ) -> dict[
+        str,
+        tuple[float, float, float],
+    ]:
+        """Get the features of all parts."""
+        features = {}
+        for obj in self.objects:
+            if obj.name.startswith("part"):
+                if obj.type == Geom3DCuboidType:
+                    features[obj.name] = self.get_object_half_extents(obj.name)
+                elif obj.type == Geom3DTriangleType:
+                    features[obj.name] = self.get_object_triangle_features(obj.name)[:3]
+                else:
+                    raise ValueError(f"Unsupported part type: {obj.type}")
+        return features
+
+    @property
+    def available_parts(self) -> list[str]:
+        """Get the names of all parts that are not currently grasped and not already
+        placed on rack."""
+        available_parts = []
+        for obj in self.objects:
+            if obj.name.startswith("part"):
+                if self.get(obj, "grasp_active") < 0.5 and not is_on_top(
+                    self.rack_pose,
+                    self.rack_half_extents,
+                    self.get_object_pose(obj.name),
+                    self.get_object_half_extents(obj.name),
+                ):
+                    available_parts.append(obj.name)
+        return available_parts
+
+    @property
+    def target_object(self) -> Object | None:
```

Comment:
is this only used for unit testing? if so, let's just move it into the unit tests. for a moment I was confused about why there would only be one target object in this environment